src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,214 @@
+                WriteEvent(ResolutionStartEventId, hostNameOrAddress);
+                return ValueStopwatch.StartNew();
+            }
+
+            return default;
+        }
+
+        [NonEvent]
+        public ValueStopwatch ResolutionStart(IPAddress address)
+        {
+            Debug.Assert(address != null);
+
+            if (IsEnabled())
+            {
+                Interlocked.Increment(ref _lookupsRequested);
+                WriteEvent(ResolutionStartEventId, FormatIPAddressNullTerminated(address, stackalloc char[MaxIPFormattedLength]));
+                return ValueStopwatch.StartNew();
+            }
+
+            return default;
+        }
+
+        [NonEvent]
+        public void AfterResolution(string hostNameOrAddress, ValueStopwatch stopwatch, bool successful)
+        {
+            if (stopwatch.IsActive)
+            {
+                double duration = stopwatch.GetElapsedTime().TotalMilliseconds;
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                if (successful)
+                {
+                    ResolutionSuccess(hostNameOrAddress, duration);
+                }
+                else
+                {
+                    ResolutionFailure(hostNameOrAddress, duration);
+                }
+            }
+        }
+
+        [NonEvent]
+        public void AfterResolution(IPAddress address, ValueStopwatch stopwatch, bool successful)
+        {
+            if (stopwatch.IsActive)
+            {
+                double duration = stopwatch.GetElapsedTime().TotalMilliseconds;
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                WriteEvent(
+                    successful ? ResolutionSuccessEventId : ResolutionFailureEventId,
+                    FormatIPAddressNullTerminated(address, stackalloc char[MaxIPFormattedLength]),
+                    duration);
+            }
+        }
+
+
+        [Event(ResolutionSuccessEventId, Level = EventLevel.Informational)]
+        private void ResolutionSuccess(string hostNameOrAddress, double duration) => WriteEvent(ResolutionSuccessEventId, hostNameOrAddress, duration);
+
+        [Event(ResolutionFailureEventId, Level = EventLevel.Informational)]
+        private void ResolutionFailure(string hostNameOrAddress, double duration) => WriteEvent(ResolutionFailureEventId, hostNameOrAddress, duration);
+
+
+        [NonEvent]
+        private static Span<char> FormatIPAddressNullTerminated(IPAddress address, Span<char> destination)
+        {
+            Debug.Assert(address != null);
+
+            bool success = address.TryFormat(destination, out int charsWritten);
+            Debug.Assert(success);
+
+            Debug.Assert(charsWritten < destination.Length);
+            destination[charsWritten] = '\0';
+
+            return destination.Slice(0, charsWritten + 1);
+        }
+
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, Span<char> arg1)
```

Comment:
These overloads to `WriteEvent` should be safe as long as they are called from an `[Event]` method and the args passed to them match exactly, the signature of the method that calls them.  If they are called from elsewhere or the signature doesn't match, there is no guarantee that the result will be consumable.

Reply:
I added a comment about the Span `WriteEvent` overloads mimicking the behavior of string ones.